lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1098 @@
+    fn disconnect_socket(&mut self) {
+        let _ = self.reader_cmd_tx.try_send(ReaderCommand::Shutdown);
+        let _ = self.writer_cmd_tx.try_send(WriterCommand::Shutdown);
+        let _ = self.tcp_disconnector.shutdown();
+    }
+}
+
+/// The states that the Reader can be in.
+enum ReaderState {
+    /// Ready state; Reader is blocked on read().
+    Reading,
+    /// Reading is paused; Reader is blocked on recv().
+    Paused,
+    /// Reader will shut down in the next iteration of the run() event loop.
+    ShuttingDown,
+}
+
+/// An actor that synchronously handles the read() events emitted by the socket.
+struct Reader<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    inner: TcpReader,
+    peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+    descriptor: SyncSocketDescriptor,
+    reader_cmd_rx: Receiver<ReaderCommand>,
+    state: ReaderState,
+}
+impl<CMH, RMH, L, UMH> Reader<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    fn new(
+        reader: TcpReader,
+        peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+        descriptor: SyncSocketDescriptor,
+        reader_cmd_rx: Receiver<ReaderCommand>,
+    ) -> Self {
+        Self {
+            inner: reader,
+            peer_manager,
+            descriptor,
+            reader_cmd_rx,
+            state: ReaderState::Reading,
+        }
+    }
+
+    /// Handle read events, or wait for the next `ReaderCommand` if reads are
+    /// paused. This implementation avoids busy loops and lets the thread go to
+    /// sleep whenever reads or channel commands are pending.
+    fn run(&mut self) {
+        use ReaderState::*;
+
+        // 8KB is nice and big but also should never cause any issues with stack
+        // overflowing.
+        let mut buf = [0; 8192];
+
+        loop {
+            self.do_try_recv();
+
+            match self.state {
+                Reading => self.do_read(&mut buf),
+                Paused => self.do_recv(),
+                ShuttingDown => break,
+            };
+        }
+
+        // Shut down the underlying stream. It's fine if it was already closed.
+        let _ = self.inner.shutdown();
+        // Send a signal to the Writer to do the same.
+        let _ = self
+            .descriptor
+            .writer_cmd_tx
+            .try_send(WriterCommand::Shutdown);
+    }
+
+    /// Checks for a command in a non-blocking manner, handling the command
+    /// if there was one.
+    fn do_try_recv(&mut self) {
+        match self.reader_cmd_rx.try_recv() {
```

Comment:
I tried implementing this, but ran into the issue that when reading is paused, the `Reader` _needs_ something else to block on, otherwise it requires a busy loop on `should_read.load()` in order to detect that it should resume reads.

Reply:
On it. Starting to think that we can indeed get rid of the entirety of the `crossbeam` dependency